platform/tls/generic/mbedtls/src/mender-tls.c

```diff
@@ -127,77 +127,95 @@ mender_tls_init(void) {
         if (MENDER_OK != mender_storage_delete_authentication_keys()) {
             mender_log_warning("Unable to delete authentication keys");
         }
     }
 
-    /* Buffer used for user-provided key */
-    char  *user_provided_key        = NULL;
-    size_t user_provided_key_length = 0;
+    /* Get user provided key (callback is optional) */
+    if (NULL != get_user_provided_keys) {
+        char  *user_provided_key        = NULL;
+        size_t user_provided_key_length = 0;
 
-    if (MENDER_OK != (ret = get_user_provided_keys(&user_provided_key, &user_provided_key_length))) {
-        mender_log_error("Unable to get user provided key");
-        goto END;
-    }
-    if (NULL != user_provided_key) {
-        mender_log_info("Getting authentication key...");
-        if (MENDER_OK
-            != (ret = mender_tls_get_authentication_keys(&mender_tls_private_key,
-                                                         &mender_tls_private_key_length,
-                                                         &mender_tls_public_key,
-                                                         &mender_tls_public_key_length,
-                                                         user_provided_key,
-                                                         user_provided_key_length))) {
+        mender_log_debug("Retrieving user provided authentication keys");
+
+        if (MENDER_OK != get_user_provided_keys(&user_provided_key, &user_provided_key_length)) {
             mender_log_error("Unable to get user provided authentication key");
-            goto END;
+            return MENDER_FAIL;
         }
-        /* Retrieve or generate private and public keys */
-    } else if (MENDER_OK
-               != (ret = mender_storage_get_authentication_keys(
-                       &mender_tls_private_key, &mender_tls_private_key_length, &mender_tls_public_key, &mender_tls_public_key_length))) {
-        /* Generate authentication keys */
-        mender_log_info("Generating authentication keys...");
-        if (MENDER_OK
-            != (ret = mender_tls_get_authentication_keys(
-                    &mender_tls_private_key, &mender_tls_private_key_length, &mender_tls_public_key, &mender_tls_public_key_length, NULL, 0))) {
-            mender_log_error("Unable to generate authentication keys");
-            goto END;
+
+        if ((NULL == user_provided_key) || (0 == user_provided_key_length)) {
+            mender_log_error("User provided authentication key is empty");
+            return MENDER_FAIL;
         }
 
-        /* Record keys */
         if (MENDER_OK
-            != (ret = mender_storage_set_authentication_keys(
-                    mender_tls_private_key, mender_tls_private_key_length, mender_tls_public_key, mender_tls_public_key_length))) {
-            mender_log_error("Unable to record authentication keys");
-            goto END;
+            != mender_tls_get_authentication_keys(&mender_tls_private_key,
```

Comment:
I think we should split `mender_tls_get_authentication_keys` in two different functions. The API here is very obscure: if you pass user keys it basically converts the format and extracts public key, while if these are null then it generates keys.

We should instead refactor the internals of it, kill this function, and instead us `mender_tls_user_provided_authentication_keys` and `mender_tls_generate_authentication_keys`

Reply:
Okey, I'll try to do this in a separate PR ;)